Tutorials: Update Next.js pages router (#12442)
PostHog/posthog.com: contents/tutorials/nextjs-analytics.md

---
title: 'How to set up Next.js analytics, feature flags, and more'
date: 2025-08-13
author:
  - ian-vanagas
showTitle: true
sidebar: Docs
tags:
  - configuration
  - feature flags
  - persons
  - events
---

Next.js is one of the web's most popular frameworks. Built on React, it provides optimizations and abstractions to help developers build fast and performant apps and sites. 

To make sure you get the most out of Next.js, you can use PostHog to track events, identify users, use feature flags, and more. To help you get started, this tutorial walks you through:

> 1. Building a basic Next.js (pages router) blog with user authentication

Reviewer: Should we specify in the title?

I think today, most new projects are app router (the direction vercel is pushing for. I know we have a separate post

Author: Yeah, I think I'm going to merge this then do another PR to swap them and redirect the app-router URL.

2. Adding PostHog to it
3. Setting up the features of PostHog like custom event capture, user identification, and feature flags

import { CalloutBox } from 'components/Docs/CalloutBox'

<CalloutBox type="info" title="Looking for the app router?">

  If you use Next.js with the **app router**, check out our other [Next.js app router analytics tutorial](/tutorials/nextjs-app-directory-analytics).
    
</CalloutBox>

## Creating our Next.js app

> First, [install Node](https://nodejs.dev/en/learn/how-to-install-nodejs/) (18.18 or newer) and then run:

Reviewer: should we just say node 20? node 18 is eol at this point and I've had more and more trouble with it lately

Author: This is what Vercel says 🤷

})
```

Next, get these details from GitHub by going to [developer settings](https://github.com/settings/developers). Click **New OAuth App**, add a name, set the homepage URL to `http://localhost:3000`, the authorization callback URL to `http://localhost:3000/api/auth/callback/github`, and then click **Register application**.

Create a new OAuth app and get the client ID and client secret.

![GitHub](https://res.cloudinary.com/dmukukwp6/image/upload/w_1600,c_limit,q_auto,f_auto/Clean_Shot_2025_08_13_at_10_08_00_2x_d38cb2039c.png)

Copy the client ID and generate and copy a new client secret. With these, create a `.env.local` file in the main app (`pages-tutorial`) folder and set them as `GITHUB_ID` and `GITHUB_SECRET`. You'll also need to add a `NEXTAUTH_URL` ([`http://localhost:3000`](http://localhost:3000/) for now) and `NEXTAUTH_SECRET` (which you can generate on [this site](https://generate-secret.vercel.app/32) or by creating a random 32-character string) value.

```bash
GITHUB_ID=<github_client_id>
GITHUB_SECRET=<github_client_secret>
NEXTAUTH_URL=http://localhost:3000
NEXTAUTH_SECRET=<random_32_character_string>
```

### Setting up sessions

> With NextAuth and GitHub set up, we now have the infrastructure to authenticate users. Now, we can implement user sessions to let them log in and out as well as get their details. 

Reviewer: ```suggestion
With NextAuth and GitHub set up, we have the infrastructure to authenticate users. Now, we can implement user sessions to let them log in and out as well as get their details. 
```

Minor nit: two nows close together reads funny

The first step to doing this is adding a `SessionProvider` from `next-auth/react` to `_app.js` like this:

```js
// pages/_app.js
import "@/styles/globals.css";
import { SessionProvider } from "next-auth/react"

export default function App({ Component, pageProps: { session, ...pageProps } }) {
  return (
    <SessionProvider session={session}>
      <Component {...pageProps} />
    </SessionProvider>
  );
}
```

Next, add the session details, and the ability to sign in and out to our `index.js` page. We can do this with the NextAuth `useSession()` hook as well as its methods for signing in and out.

We set up a check to see if there is a session and show details about the user and a button to sign out. If there isn’t a session, we show a button to sign in. Together, it looks like this:

> ```js

Reviewer: I think this is editing an existing file. Consider using the `// +` comment trailing a line to highlight lines added here. Might make it easier to follow in longer tutorials like this.

When you click sign in, you go through the sign in flow with GitHub, and get redirected back to the app with an active session.

![Signed in](https://res.cloudinary.com/dmukukwp6/image/upload/q_auto,f_auto/Clean_Shot_2025_08_13_at_10_18_27_2x_67dab4aa67.png)

Once this is working, we have all the functionality we want in our Next.js app and it’s time to add PostHog.

## Adding PostHog

> At this point, you need a PostHog instance (it's [free to sign up](https://app.posthog.com/signup)). Once created, get your project API key and instance address from [your project settings](https://us.posthog.com/settings/project) and add it to your `.env.local` file.

Reviewer: ```suggestion
At this point, you need a PostHog project (it's [free to sign up](https://app.posthog.com/signup)). Once created, get your project API key and instance address from [your project settings](https://us.posthog.com/settings/project) and add it to your `.env.local` file.
```

Maybe don't use instance. They might try self-hosting lol

//...
```

> Go to a post, click **Like**, and then check your PostHog instance to see the custom event show up.

Reviewer: ```suggestion
Go to a post, click **Like**, and then check your PostHog project's [activity tab](https://app.posthog.com/activity/explore) to see the custom event show up.
```
